### PR TITLE
COMMERCE-10535 Enable customised cookie configuration modal for product comparison

### PR DESCRIPTION
**Pull request by dsanz**
### References

- https://issues.liferay.com/browse/COMMERCE-10535
- This PR supersedes https://github.com/liferay-frontend/liferay-portal/pull/2932

### What is the goal of this PR?

To make product comparison feature to obey user consent for `functional` cookies. This feature requires a non-essential cookie which must be accepted by users.

In doing so, we added a DXP JS api to handle cases like this in the future, so that other applications don't need to reinvent the wheel when some feature requires cookie acceptance

### What does it look like?

[cookie-banner.webm](https://user-images.githubusercontent.com/861014/210221371-b1053c37-a2aa-4eda-abf7-7f39dbac264f.webm)

### Steps to reproduce

1. Enable cookie handling feature by adding `feature.flag.LPS-142518=true` to your portal-ext.properties and restarting the server, or just add that property dynamically via server admin panel
2. Configure instance-level cookie preferences: browse to **Instance Settings → Cookies → Preference Handling** and check **Enabled**
3. You'll see the cookie banner. Click on **Configuration** link and reject all cookies
4. Create a site with the `minium` initializer
5. Browse to the site catalog page (/web/site-name/catalog)
6. Select a product by clicking on the checkbox at the top-left corner in the product card

Before this change, it was possible to select products for comparison, and the minicompare component was rendered. Also, the `COMMERCE_COMPARE_cpDefinitionIds_*` cookie was created and stored in the user's device.

After this change, the product selection process ensures `functional` cookies are accepted before setting `COMMERCE_COMPARE_cpDefinitionIds_*` cookie. Also, product comparison widget is not displayed until user these cookies are accepted. Cookie consent modal is shown to allow user to give consent.

### Other notes

If you want to repeat the test, you can:

1. Open the devtools panel, then delete the `COMMERCE_COMPARE_cpDefinitionIds_*` cookie and set `CONSENT_TYPE_FUNCTIONAL` cookie to `false`
2. Reload the page

Then, you can just go to step 5 in the reproduction steps.

### CODE REVIEW

File: modules/apps/cookies/cookies-banner-web/src/main/resources/META-INF/resources/cookies_banner/js/CookiesBanner.js
Comment (dsanz) [inline modules/apps/cookies/cookies-banner-web/src/main/resources/META-INF/resources/cookies_banner/js/CookiesBanner.js:209]: @bryceosterhaus @kresimir-coko I've added this thing as a new API so that dxp apps don't need to implement complex logic to handle the cookie banner interaction. 

In this case I'm using promises to wait for the "right" user input. I could not find much cases like this in the JS apis we provide, so I'd like to hear your thoughts

**Comment by dsanz** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js
Comment (dsanz) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js:119]: As this funcion component delegates rendering to `checkCookieConsentForTypes()`, there's no output to return. This pattern looks a bit hacky to me, I could not find a better way to integrate the API into the react cycle. @kresimir-coko @bryceosterhaus maybe there's a better way?

**Comment by dsanz** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js
Comment (bryceosterhaus) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js:119]: Just so I fully understand, you want to make sure that `checkCookieConsentForTypes` is confirmed before rending the main content of this component?

Comment (bryceosterhaus) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js:119]: If they deny the cookie, does `setFunctionalCookiesConsent(true);` still get set?

**Review by bryceosterhaus — COMMENTED** [review]
I think your idea makes sense, although it _feels_(very subjective) like it takes a a bit of energy to think through. let me see if I can pull this down and re-work it slightly.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js
Comment (dsanz) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js:119]: yes, we need functional cookies to be accepted before rendering the main content. I'm taking this as a kind of "loading..." state while fetched data comes, with the peculiarity that in some cases there's no need to run any real fetch as the data is already there.

> If they deny the cookie, does setFunctionalCookiesConsent(true); still get set?

It should not as  `setFunctionalCookiesConsent(true)` is in the resolve() part. However, there are a remote chance of not syncing the component state with the cookie status as we do in the initial `useEffect()` call. Maybe we can use:

`setFunctionalCookiesConsent(checkConsent(COOKIE_TYPES.FUNCTIONAL))`

instead

Comment (dsanz) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js:119]: I'm starting to think that some sort of `useCookieConsent()` hook might make sense.

Nevertheless, as a side note, we can do that after sending this PR as there's a tight deadline for the cookie management feature (which depends on this product comparison changes) to be released

**Review by edalgrin — APPROVED** [review]
Since we need to merge this code ASAP I would say LGTM!! If any optimization is missing we can work on that later.

**Comment by dsanz** [comment]
ci:test:relevant

**Comment by dsanz** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js
Comment (dsanz) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_compare/MiniCompare.js:99]: That condition looks also part of the "contract" between the app and the consent modal, because it short-circuits regular component rendering as well as the behavior of the effect.

So, basically, we'd need to abstract something out around these lines:

- A list of cookie types that need to be approved
- Some customizations for the cookie modal 
  - (title, alert box)
  - Success and failure toast alert texts/titles after modal is closed
- Extra, app-level conditions that must be met to check for consent instead of rendering the component (in this PR, `items > 0`). In addition, contract would ensure cookie types are approved as well.

**Comment by dsanz** [comment]
@john-co @mateusralv there seems to be an error in the relevant suite related to the build process itself. Could you please help us to check? Thank you

**Comment by john-co** [comment]
cc/ @eltonccdantas please take a look to see if any related issues

**Comment by eltonccdantas** [comment]
Sure, I'll take a look at it.

**Comment by eltonccdantas** [comment]
Hey, according to this [comparison](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=2&testrayRunIdB=2147869475&testrayRunIdA=2147731740&statusB=3), the fails look unrelated from a QA perspective. It seems safe to forward manually if no other dev review is needed. 

CC/ @mateusralv , @john-co

**Comment by victoroscuro** [comment]
ci:test:commerce

**Comment by victoroscuro** [comment]
ci:test:relevant

**Comment by dsanz** [comment]
just force-pushed to update from master

**Comment by dsanz** [comment]
ci:test:relevant

**Comment by dsanz** [comment]
ci:test:commerce

**Comment by dsanz** [comment]
ci:test:relevant

**Comment by dsanz** [comment]
ci:test:commerce

**Comment by dsanz** [comment]
@mateusralv could you please help me to verify whether relevant test suite failures are related? Thank you

**Comment by mateusralv** [comment]
I'm taking a look 🕵🏽

**Comment by mateusralv** [comment]
Hi @dsanz I took a look at the latest results of the ci:test:relevant (https://github.com/liferay-frontend/liferay-portal/pull/2938#issuecomment-1377154423) and [ci:test:commerce](https://github.com/liferay-frontend/liferay-portal/pull/2938#issuecomment-1377035991).

The failures of the relevant is not related to your PR, has a unique failure but is flaky. 🟢 

The most failures at commerce routine is a crash in gradle's daemon:
`Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)`  🟠 

Tests related to this failure:
1. com.liferay.commerce.notification.service.persistence.test.CommerceNotificationQueueEntryPersistenceTest
2. com.liferay.commerce.notification.service.persistence.test.CommerceNotificationTemplateCommerceAccountGroupRelPersistenceTest
3. com.liferay.commerce.notification.service.persistence.test.CommerceNotificationTemplatePersistenceTest
4. com.liferay.commerce.notification.test.CommerceNotificationTest
5. com.liferay.commerce.notification.test.CommerceOrderStatusNotificationTest
6. com.liferay.commerce.notification.test.CommerceSubscriptionsNotificationTest
7. com.liferay.commerce.order.rule.internal.validator.test.CORCommerceOrderValidatorTest
8. com.liferay.commerce.order.rule.service.persistence.test.COREntryPersistenceTest
9. com.liferay.commerce.order.rule.service.persistence.test.COREntryRelPersistenceTest
10. com.liferay.commerce.payment.engine.test.CommercePaymentEngineTest
11. com.liferay.commerce.payment.service.persistence.test.CommercePaymentMethodGroupRelPersistenceTest
12. com.liferay.commerce.payment.service.persistence.test.CommercePaymentMethodGroupRelQualifierPersistenceTest
13. com.liferay.commerce.price.list.change.tracking.test.CommercePriceEntryTableReferenceDefinitionTest
14. com.liferay.commerce.price.list.change.tracking.test.CommercePriceListAccountRelTableReferenceDefinitionTest
15. com.liferay.commerce.price.list.change.tracking.test.CommercePriceListChannelRelTableReferenceDefinitionTest
16. com.liferay.commerce.price.list.change.tracking.test.CommercePriceListCommerceAccountGroupRelTableReferenceDefinitionTest
17. com.liferay.commerce.price.list.change.tracking.test.CommercePriceListDiscountRelTableReferenceDefinitionTest
18. com.liferay.commerce.price.list.change.tracking.test.CommercePriceListOrderTypeRelTableReferenceDefinitionTest
19. com.liferay.commerce.price.list.change.tracking.test.CommercePriceListTableReferenceDefinitionTest
20. com.liferay.commerce.price.list.change.tracking.test.CommerceTierPriceEntryTableReferenceDefinitionTest

I'm not sure if this is related 🤔 so I run the routine again. What do you think?

**Comment by mateusralv** [comment]
ci:test:commerce

**Comment by dsanz** [comment]
Thanks for the checks @mateusralv . Commerce side is ok as per @victoroscuro comments in slack so this is safe to forward.

Please see https://github.com/brianchandotcom/liferay-portal/pull/128545

**Pull request closed without merging**
